application/src/main/java/org/togetherjava/tjbot/commands/moderation/temp/TemporaryModerationRoutine.java

```diff
@@ -138,16 +140,34 @@ private void revokeAction(@NotNull RevocationGroupIdentifier groupIdentifier) {
         return action.revokeAction(guild, target, reason);
     }
 
-    private void handleFailure(@NotNull Throwable failure,
-            @NotNull RevocationGroupIdentifier groupIdentifier) {
-        if (getRevocableActionByType(groupIdentifier.type).handleRevokeFailure(failure,
-                groupIdentifier.targetId) == RevocableModerationAction.FailureIdentification.KNOWN) {
-            return;
-        }
-
-        logger.warn(
+    private void handleFailure(@NotNull Throwable failure, long targetId,
+            @NotNull String actionName) {
+        Runnable logUnknownFailure = () -> logger.warn(
                 "Attempted to revoke a temporary moderation action for user '{}' but something unexpected went wrong.",
-                groupIdentifier.targetId, failure);
+                targetId, failure);
+
+        if (failure instanceof ErrorResponseException errorResponseException) {
+            switch (errorResponseException.getErrorResponse()) {
+                case UNKNOWN_USER -> logger.debug(
+                        "Attempted to revoke a temporary {} but user '{}' does not exist anymore.",
+                        actionName, targetId);
+                case UNKNOWN_MEMBER -> logger.debug(
+                        "Attempted to revoke a temporary {} but user '{}' is not a member of the guild anymore.",
+                        actionName, targetId);
+                case UNKNOWN_ROLE -> logger.warn(
+                        "Attempted to revoke a temporary {} but the {} role can not be found.",
+                        actionName, actionName);
+                case MISSING_PERMISSIONS -> logger.warn(
+                        "Attempted to revoke a temporary {} but the bot lacks permission.",
+                        actionName);
+                case UNKNOWN_BAN -> logger.debug(
+                        "Attempted to revoke a temporary {} but the user '{}' is not banned anymore.",
+                        actionName, targetId);
+                default -> logUnknownFailure.run();
```

Comment:
This is a design failure. Now you have the "parent class" requiring knowledge about its implementation. I.e. `Animal` having specific code for `Dog` and `Cat`. As in, this class now assumes its implementations are BAN and role-based actions. Not good.

If you want to have role-based actions, then they have to be in a specific class or method that deals only with role based actions, and then be given/picked by the subclass. Not the parent.

I am voting for going back to the idea of adding a utility class, which the implementations will then pick.

Reply:
Should've reloaded before my review, I do agree with Zabu.
Original intention was an utility method, and I would prefer that over this